src/s3ql/backends/backblaze.py

```diff
@@ -0,0 +1,704 @@
+    """An BackBlaze object open for writing"""
+
+    def __init__(self, key, backend, meta):
+        self.key = key
+        self.meta = meta
+        self.closed = False
+        self.obj_size = 0
+        self.backend = backend
+
+        # According to http://docs.python.org/3/library/functions.html#open
+        # the buffer size is typically ~8 kB. We process data in much
+        # larger chunks, so buffering would only hurt performance.
+        self.fh = tempfile.TemporaryFile(buffering=0)
+
+        self.sha1 = hashlib.sha1()
+
+    def write(self, buf):
+        self.fh.write(buf)
+        self.obj_size += len(buf)
+        self.sha1.update(buf)
+
+    def is_temp_failure(self, exc):
+        if isinstance(exc, NoSuchObject):
+            return True
+        return self.backend.is_temp_failure(exc)
+
+    @retry
+    def close(self):
+        """Close object and upload data"""
+        log.debug('started with %s', self.key)
+
+        if self.closed:
+            # still call fh.close, may have generated an error before
+            self.fh.close()
+            return
+
+        self.fh.seek(0)
+        upload_auth_token, upload_url = self.backend._get_upload_url()
+        upload_url = urllib.parse.urlparse(upload_url)
+
+        with HTTPConnection(upload_url.hostname, 443, ssl_context=self.backend.ssl_context) as conn_up:
+
+            headers = CaseInsensitiveDict()
+            headers['X-Bz-File-Name'] = self.backend.prefix + self.key
+            headers['Content-Type'] = 'application/octet-stream'
+            headers['Content-Length'] = self.obj_size
+            headers['X-Bz-Content-Sha1'] = self.sha1.hexdigest()
+
+            if self.meta is None:
+                self.meta = dict()
+            self.backend._add_meta_headers(headers, self.meta)
+
+            self.backend._do_request('POST', upload_url.path + '?' + upload_url.query,
+                                     conn_up,
+                                     headers=headers,
+                                     body=self.fh,
+                                     auth_token=upload_auth_token,
+                                     body_size=self.obj_size)
+
+        self.fh.close()
+        self.closed = True
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *a):
+        self.close()
+        return False
+
+    def get_obj_size(self):
+        if not self.closed:
+            raise RuntimeError('Object must be closed first.')
+        return self.obj_size
+
+
+class AuthenticationExpired(Exception):
+    """Raised if the provided Authentication Token has expired"""
+
+    def __init__(self, msg):
+        super().__init__()
+        self.msg = msg
+
+    def __str__(self):
+        return 'Auth token expired. Server said: %s' % self.msg
+
+def _parse_retry_after(header):
+    """Parse headers for Retry-After value"""
+
+    hit = re.match(r'^\s*([0-9]+)\s*$', header)
+    if hit:
+        val = int(header)
+    else:
+        val = 1
```

Comment:
```
try:
  val = int(header)
except ValueError:
   int = 1
```
would be a  lot clearer. However, just like s3c you should log an error if there's an invalid value returned by the server.